### ISSUE TITLE
Add compatibility test between releases

### ISSUE THREAD

**Issue by t-mat** [Cyan4973/lz4#103]
Recently, [Joffrey LABORDE (implicitly) pointed out](https://groups.google.com/d/msg/lz4c/D_hSD19ykqU/DE2eecLIrYcJ), there is no clear evidence for interoperability between releases.  Also there is a similar FAQ : "Are 64bit (lz4c) and 32bit (lz4c32) interoperable ?"
The following Python3 script is initial answer for these questions :

 - https://gist.github.com/t-mat/3c5e1e4df3b250691caf

This automated test compress and decompress between

 - All git tagged releases
 - `lz4c` and `lz4c32`
 - `-1` and `-9`

I hope this test will guarantee basic backward and forward compatibility for all LZ4 releases.

### Usage

```sh
cd
mkdir lz4-test
cd lz4-test
curl -O https://gist.githubusercontent.com/t-mat/3c5e1e4df3b250691caf/raw/test-lz4-releases.py
chmod +x ./test-lz4-releases.py
./test-lz4-releases.py
```
[edit] : Add `chmod` 

### Details

 - Checkout all tagged release code (pattern `r[0-9][0-9][0-9]`)
    - Compiles `lz4c` and `lz4c32`
    - Rename these executables. Add tag as suffix to these executables (eg `lz4c32.r129`)
    - Copy all executables to grand parent directory
 - Copy `README.md` as test file (`test.dat`)
 - Compress `test.dat` by all executables
   - Use `lz4c.<TAGS>` and `lz4c32.<TAGS>`
   - Specify `-1zf` and `-9zf`
 - Remove dupilcated `.lz4` files
 - Decompress all `.lz4` files by all executables
   - Compare decompressed data with `test.dat`
   - Remove the file if there is no difference
   - Report the error if difference is found

### Remained Problems

 - More test data is needed for serious test.
 - Test must be running on TravisCI.
    - Also VC++ and appveyor ?
 - Dependency to python is not fit for LZ4 project.
 - Is `make cc=clang` needed ?

**Comment by Cyan4973** [comment]
Hi Takayuki

Testing 32-bits <-> 64-bits compatibility is already part of the CI test suite. This is probably not clear enough if you have missed it.
Look into [/program/Makefile](https://github.com/Cyan4973/lz4/blob/master/programs/Makefile), and the paragraph [test-lz4c32](https://github.com/Cyan4973/lz4/blob/master/programs/Makefile#L227), producing for example the [following result](https://travis-ci.org/Cyan4973/lz4/jobs/62204572).

It tests several file sizes and compression modes.

Now, maybe what could be done is to expand on it : 
 - Label more clearly some of these tests as "32/64 bits interoperability tests"
 - Only compress 32-bits -> decompress 64-bits is tested. Maybe the reverse operation should be added.
 - Adding more compression cases is also possible : `datagen` can be used to create various fils sizes with various compressibility distribution, from highly compressible, to barely or impossible to compress. Current tests don't try different distributions.
 - And potentially, add a visible note somewhere that this interoperability is guaranteed and tested.

**Comment by t-mat** [comment]
I'm sorry for my ambiguous text !  Also I'm completely overlooked to mention to previous work such as `test-lz4c32`.

> I suspect it's possible to use shell commands to download precise releases.

I think it's possible but code may be cryptic :scream: also generally script languages are more platform independent than shell script.
Do you have any favorite script language ?  Since this script itself is simple, I'll try to rewrite it in your favorite language.

**Comment by Cyan4973** [comment]
> generally script languages are more platform independent than shell script.

That's a good enough reason.

> Do you have any favorite script language ?

Not at all. 
And I don't see any other script language with better reach & support than Python.
So it's fine with me.

**Comment by t-mat** [comment]
Thanks.  I have one more question.  Where should I put testing script ?

Aside from scripts location, I would like to implement the following things

 - [ ] TravisCI Integration
 - [ ] Include `dev` branch to the test
 - [ ] Show only unique `.lz4` files like this:
```
test_dat_1_32_r116.lz4  2279
test_dat_1_32_r129.lz4  2279
test_dat_1_64_r129.lz4  2305
test_dat_9_32_r116.lz4  2125
test_dat_9_32_r124.lz4  2157
test_dat_9_32_r129.lz4  2125
```

Also I leave the following things for now

 - Test case for clang, VC++ and appveyor
 - More complete/realistic test data

Since it's beginning of forward/backward compatibility test, start with simpler condition and firstly improving test flow itself looks good decision.

Please let me know anything important thing I've overlooked.

**Comment by Cyan4973** [comment]
Travis CI integration looks the most important feature here.

I'm not sure if it is useful to target a "dev" branch specifically : the dev branch will be tested by Travis automatically every time an update is done to it. So it looks redundant. I would keep your initial idea, in combination with Travis it will test the currently modified branch with previously published releases.

I don't exactly understand the 3rd item.

**Comment by t-mat** [comment]
> I'm not sure if it is useful to target a "dev" branch specifically 

Yes, you are right.  It's wrong word.  It should be latest commit (HEAD).

> I don't exactly understand the 3rd item.

I believe this log will be useful to indicate unintended change.  Suppose harmless looking (it does not change any format related thing) commit/PR will make the following CI log :

```
test_dat_1_32_r116.lz4  2279
test_dat_1_32_r129.lz4  2279
test_dat_1_64_r129.lz4  2305
test_dat_9_32_r116.lz4  2125
test_dat_9_32_r124.lz4  2157
test_dat_9_32_r129.lz4  2125
test_dat_9_32_HEAD.lz4  2125  <-- suspicious result
```

Here, HEAD means latest commits result.  Since HEAD is harmless, it's output should be same as previous master, and same file must be removed.
But it remains.  So, this log indicates something has been changed, maybe accidentally.

I think this is useful for PR senders and reviewers.

**Comment by Cyan4973** [comment]
okay.
Indeed, they are many subtle changes that can affect the output of the compression function in a minor way, and without affecting its format compliance. Most of the time, it does not really matter.

But maybe what you mean is that sometimes, we don't intend to make any change at all to the output. So, if change there is, your test will help detect that. So it's okay.

**Comment by Cyan4973** [comment]
Solved by @t-mat #104
